pandablocks/state.py

```diff
@@ -0,0 +1,105 @@
+
+    The file format is a series of PandA-client command strings
+    """
+
+    def __init__(self, host: str):
+        """
+        Args:
+            host (str): the IP address or DNS name of a PandA
+        """
+        self._host = host
+        self._client = BlockingClient(host)
+        self._client.connect()
+
+    def __del__(self):
+        self._client.close()
+
+    def _send_get(self, command: str) -> List[str]:
+        """
+        Send 'command' to host, get the results and return them as a list of strings
+
+        Args:
+            command str: PandA command on which to perform 'get'
+        """
+        lines = self._client.send(Get(command))
+
+        # send may return a single bytes or a list of bytes
+        if isinstance(lines, bytes):
+            result = [lines.decode()]
+        else:
+            result = [line.decode() for line in lines]
+        return result
+
+    def save(self) -> List[str]:
+        """
+        Create a sequence of commands that will set up a PandaA to mirror
+        the current host's state.
+
+        Returns:
+            str: A sequence of PandA-client command strings
+        """
+
+        def save_table(table: str) -> None:
+            assert table[-1] == "<", f"bad response to *CHANGES.TABLE: {table}"
+            commands.append(table + "B")
+            for line in self._send_get(table[:-1] + ".B"):
+                commands.append(line)
+            else:
+                commands.append("")
+
+        def save_metatable(meta: str) -> None:
+            commands.append(meta)
+            for line in self._send_get(meta[:-1]):
+                commands.append(line)
+            else:
+                commands.append("")
+
+        def save_metadata(line: str) -> None:
+            if line[-1] == "<":
+                save_metatable(line)
+            else:
+                commands.append(line)
+
+        # save the CONFIG state
+        commands = self._send_get("*CHANGES.ATTR")
+        commands += self._send_get("*CHANGES.CONFIG")
+        # save individual tables
+        tables = self._send_get("*CHANGES.TABLE")
+        for table in tables:
+            save_table(table)
+        # save metadata which is a mix of values and tables
+        metadata = self._send_get("*CHANGES.METADATA")
+        for line in metadata:
+            save_metadata(line)
+
+        return commands
+
+    def load(self, commands: List[str]):
+        """
+        Send a sequence of commands to host. Provide a sequence obtained from
+        State.save() in order to restore a previous state.
+
+        Args:
+            commands (str): a series of PandA-client commmand strings
+        """
+        full_command = None
+        for command_line in commands:
+            if full_command is None:
+                if is_multiline_command(command_line):
+                    full_command = [command_line]
+                else:
+                    self._client.send(Raw([command_line]))
```

Comment:
Can we check that the response is "OK", and report all the commands that weren't ok please?

Reply:
done in 04b5d23 but I'm not sure I'm happy with load - it feels like some of it needs to shift into commands.py

Reply:
see https://github.com/PandABlocks/PandABlocks-client/issues/7#issuecomment-737188682